src/Frontend/Core.hs

```diff
@@ -207,31 +287,78 @@ class Page p => FormPage p where
     guardPage :: (ActionM m) => p -> m (Maybe UriPath)
     guardPage _ = pure Nothing
 
+-- | Representation of a 'FormPage' suitable for passing to 'formPage' and generating Html from it.
+data FormPageRep p = FormPageRep (View (Html ())) ST p
 
--- | Defines some properties for pages
-class Page p where
-    isPrivatePage :: proxy p -> Bool
-    isPrivatePage _ = True
+instance (Show p) => Show (FormPageRep p) where
+    show (FormPageRep _v _a p) = show p
 
-    extraPageHeaders  :: p -> Html ()
-    extraPageHeaders _ = nil
+instance Page p => Page (FormPageRep p) where
+    isPrivatePage _ = isPrivatePage (Proxy :: Proxy p)
+    extraPageHeaders (FormPageRep _v _a p) = extraPageHeaders p
 
-    extraBodyClasses  :: p -> [ST]
-    extraBodyClasses _ = nil
+instance FormPage p => ToHtml (FormPageRep p) where
+    toHtmlRaw = toHtml
+    toHtml (FormPageRep v a p) =  toHtml $ formPage v frm p
+      where
+        frm bdy = DF.childErrorList "" v >> DF.form v a bdy
 
-instance Page () where
-    isPrivatePage _ = False
+data FormPageHandler m p = FormPageHandler
+    { _formGetPage   :: m p
+    , _formProcessor :: FormPagePayload p -> m (FormPageResult p)
+    }
 
-instance Page ST where
-    isPrivatePage _ = True -- safer default, might need to be changed if needed
+-- | (this is similar to 'formRedirectH' from "Servant.Missing".  not sure how hard is would be to
+-- move parts of it there?)
+--
+-- Note on file upload: The 'processor' argument is responsible for reading all file contents before
+-- returning a WHNF from 'popTempCsvFile'.  'cleanupTempCsvFiles' will be called from within this
+-- function as a 'processor' finalizer, so be weary of lazy IO!
+--
+-- Note that since we read (or write to) files eagerly and close them in obviously safe
+-- places (e.g., a parent thread of all potentially file-opening threads, after they all
+-- terminate), we don't need to use `resourceForkIO`, which is one of the main complexities of
+-- the `resourcet` engine and it's use pattern.
+form :: (FormPage p, Page p, ActionM m) => FormPageHandler m p -> ServerT (FormHandler p) m
+form formHandler = getH :<|> postH
+  where
+    getPage = _formGetPage formHandler
+    processor = _formProcessor formHandler
 
-instance (Page a, Page b) => Page (Beside a b) where
-    isPrivatePage _ = isPrivatePage (Proxy :: Proxy a) || isPrivatePage (Proxy :: Proxy b)
-    extraPageHeaders (Beside a b) = extraPageHeaders a <> extraPageHeaders b
+    guard page = do
+        r <- guardPage page
+        maybe (pure ()) (redirect . absoluteUriPath) r
```

Comment:
You can't, since `redirect` is not pure.
An explicit bind is shorter: `maybe (pure ()) (redirect . absoluteUriPath) =<< guardPage page`.

Then the `maybe (pure ())` is `mapM_` from `Data.Traversable`:

`mapM_ (redirect . absoluteUriPath) =<< guardPage page`